providers/inmemory/provider.go

```diff
@@ -81,13 +74,28 @@ func New(root string, db metadata.DB, logger *zerolog.Logger, agentOpts ...labag
 }
 
 func (p *provider) CreateNodeGroup(ctx context.Context, id string, cdef metadata.ClusterDefinition) (*p2plab.NodeGroup, error) {
-	var ns []metadata.Node
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	numPorts := 0
+	for _, group := range cdef.Groups {
+		numPorts += group.Size
+	}
+
+	freePorts, err := freeport.GetFreePorts(numPorts * 2)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
We grab the freeports in one call here instead of the `portHelper`.

Reply:
good idea :+1: